etl/steps/data/garden/health/2023-04-25/wgm_2018.py

```diff
@@ -0,0 +1,353 @@
+    gender_missing = set(set(MAPPING_GENDER_VALUES) | {"all"}).difference(df["gender"])
+    if gender_unexpected:
+        raise ValueError(f"Unexpected gender ID {gender_unexpected}")
+    if gender_missing:
+        raise ValueError(f"Missing gender ID {gender_missing}")
+
+
+def _make_df_with_share_answers(df: pd.DataFrame, weight_column: str = "weight_intra_country") -> pd.DataFrame:
+    # 1. broken down by gender and age group
+    df_gender_age = _make_individual_df_with_share_answers(
+        df, dimensions=["gender", "age_group"], weight_column=weight_column
+    )
+    # 2. broken down by gender
+    df_gender = _make_individual_df_with_share_answers(df, dimensions=["gender"], weight_column=weight_column)
+    # 3. broken down by age_group
+    df_age = _make_individual_df_with_share_answers(df, dimensions=["age_group"], weight_column=weight_column)
+    # 4. no breakdown
+    df_nb = _make_individual_df_with_share_answers(df, weight_column=weight_column)
+    # Combine dataframes
+    log.info("wgm_2018: combining dataframes into combined one")
+    df_combined = pd.concat([df_nb, df_gender, df_age, df_gender_age], ignore_index=True)
+    # Sanity check
+    x = df_combined.groupby(["country", "year", "question", "gender", "age_group"], observed=True)[["share"]].sum()
+    assert x[
+        abs(x.share - 100) > 0.1
+    ].empty, "The share was not correctly estimated! Sum of shares does not sum up to 100% (we allow for 0.1% error)"
+    return df_combined
+
+
+def _make_individual_df_with_share_answers(
+    df: pd.DataFrame, weight_column: str, dimensions: List[str] = []
+) -> pd.DataFrame:
+    """Obtain table with answer percentages and counts to each question for all demographic groups and countries.
+
+    For each question, obtain the number of each registered answer. Also, obtain the "weighted number", which is given
+    by weighting each answer according to the `weight_column` column. This is later used to obtain the share.
+
+    The `weight_column` has either the value of "weight_intra_country" (used to standardise different demographics within a country) or
+    "weight_inter_country" (to standardise a variable across different countries)
+    """
+
+    log.info(f"wgm_2018: building dataframe with share of answers for dimensions {dimensions}")
+    operations = ["sum", "count"]
+    columns_index_base = ["country", "year", "question", "answer"]
+    columns_index = columns_index_base + dimensions
+    df_ = df.groupby(columns_index, observed=True).agg({weight_column: operations})
+    df_.columns = operations
+    df_ = df_.reset_index()
+    # For each question, now obtain the percentage of each of its answers (weighted).
+    columns_normalise = [col for col in columns_index if col not in ["answer"]]
+    df_weights_sum = df_.groupby(columns_normalise, as_index=False)[["sum"]].sum()
+    df_ = df_.merge(df_weights_sum, on=columns_normalise, suffixes=("", "_denominator"))
```

Comment:
Here possibly you could also groupby transform, insted.

Reply:
totally